genesyscloud/oauth_client/data_source_genesyscloud_oauth_client.go

```diff
@@ -0,0 +1,43 @@
+package oauth_client
+
+import (
+	"context"
+	"fmt"
+	"terraform-provider-genesyscloud/genesyscloud"
+	"time"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/retry"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+)
+
+func dataSourceOAuthClientRead(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	sdkConfig := m.(*genesyscloud.ProviderMeta).ClientConfig
+	oauthClientProxy := getOAuthClientProxy(sdkConfig)
+
+	name := d.Get("name").(string)
+
+	// Find first non-deleted oauth client by name. Retry in case new oauth client is not yet indexed by search
+	return genesyscloud.WithRetries(ctx, 15*time.Second, func() *retry.RetryError {
+		clients, getErr := oauthClientProxy.getAllOAuthClients(ctx)
+		if getErr != nil {
+			return retry.NonRetryableError(fmt.Errorf("Error requesting oauth client %s: %s", name, getErr))
+		}
+
+		if len(*clients) == 0 {
+			return retry.RetryableError(fmt.Errorf("No oauth clients found with name %s", name))
+		}
+
+		for _, oauth := range *clients {
+			if oauth.Name != nil && *oauth.Name == name &&
+				oauth.State != nil && *oauth.State != "deleted" {
+				d.SetId(*oauth.Id)
+				return nil
+			}
+		}
+
+		return nil
```

Comment:
I think this should be a retryable error?

Reply:
This should be a non-retryable method because if you look at the refactored flow, you first try to see:

1.  If not data is found retry the lookup 
2.  If you got data back, search through them looking for the record.
3.  If you still cant find the record then finally return the non-retryable record.

Reply:
Done

Reply:
I'm still a bit confused. If no entities are returned - retry. If some entities are returned but none that match the name - don't retry. Isn't it possible that other oauthclients could exist, but that the new oauthclient hasn't been indexed yet? Especially since we're not filtering by name